### PR TITLE
Enable wasm for bzip2/libpng and rework libsdl

### PR DESCRIPTION
**Pull request by SirLynix**

### CODE REVIEW

**Comment by SirLynix** [comment]
windows arm builds fine but fails because it tries to start an executable (and the virtual machine is not ARM), how to handle this?

**Comment by waruqi** [comment]
> windows arm builds fine but fails because it tries to start an executable (and the virtual machine is not ARM), how to handle this?

```lua
if not package:is_cross() then
    os.runv()
end
```

**Comment by SirLynix** [comment]
I forgot package:is_cross applied to something else than the "cross" platform.

**Comment by SirLynix** [comment]
I don't get why it can't find Xext.h on fedora.

Cross seems to be lacking system include, and Wasm can't find emscripten include despite using emcc 🤔

**Comment by waruqi** [comment]
> I don't get why it can't find Xext.h on fedora.

add_deps("libxext") and pass `packagedeps="libxext"` tp cmake.install?

> Cross seems to be lacking system include, and Wasm can't find emscripten include despite using emcc 🤔

maybe we need improve find_emsdk and return {includedirs, linkdirs}

and improve tools.cmake.install to pass them if platform is wasm.

emscripten/system/include/emscripten/eventloop.h

**Comment by SirLynix** [comment]
> add_deps("libxext") and pass packagedeps="libxext" tp cmake.install?

from my understanding this is done already, except with include dirs (as libsdl only requires includes to build)

> maybe we need improve find_emsdk and return {includedirs, linkdirs}
>
> and improve tools.cmake.install to pass them if platform is wasm.

I'm more surprised that emcc doesn't use its emscripten/system/include as an include folder automatically

**Comment by waruqi** [comment]
> add_deps("libxext") and pass packagedeps="libxext" tp cmake.install?

missing opt argument.

https://github.com/xmake-io/xmake-repo/blob/e6cebae23e71816e94554a968775ae795bae9547/packages/l/libsdl/xmake.lua#L168

**Comment by SirLynix** [comment]
oh 🤦

File: packages/l/libsdl/xmake.lua [hunk 1]
```diff
@@ -126,23 +141,10 @@ package("libsdl")
...
-        table.insert(configs, "--enable-shared=" .. (package:config("shared") and "yes" or "no"))
-        table.insert(configs, "--enable-static=" .. (package:config("shared") and "no" or "yes"))
-        if package:config("pic") ~= false then
-            table.insert(configs, "--with-pic")
-        end
+        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
+        table.insert(configs, "-DLIBTYPE=" .. (package:config("shared") and "SHARED" or "STATIC"))
         local opt
         if package:is_plat("linux") then
             local cflags = {}
```
Comment (waruqi) [inline packages/l/libsdl/xmake.lua:150]: we can also use packagedeps for linux, it's older configuration here.

Comment (SirLynix) [reply packages/l/libsdl/xmake.lua:150]: I left it that way to ensure sdl only has include dirs, as it shouldn't link theses libraries (packagedeps will add -l flag)

Comment (waruqi) [reply packages/l/libsdl/xmake.lua:150]: ok

**Comment by SirLynix** [comment]
```
/home/runner/.config/.xmake/packages/c/cmake/3.24.2/68b116f83c5c4fb1b93d0f5f7a07021e/bin/cmake -DCMAKE_BUILD_TYPE=Release -DLIBTYPE=SHARED -DCMAKE_INSTALL_PREFIX=/home/runner/.config/.xmake/packages/l/libsdl/2.26.1/be1d6dd235e646038700e08273979bb6 -DCMAKE_INSTALL_LIBDIR:PATH=lib -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/home/runner/work/xmake-repo/xmake-repo/emsdk-cache-shared/emsdk-main/fastcomp/emscripten/cmake/Modules/Platform/Emscripten.cmake -DCMAKE_POSITION_INDEPENDENT_CODE=ON /home/runner/.config/.xmake/cache/packages/2212/l/libsdl/2.26.1/source
-- CMAKE_C_IMPLICIT_INCLUDE_DIRECTORIES= 
-- CMAKE_CXX_IMPLICIT_INCLUDE_DIRECTORIES= 
```

It's weird because Emscripten.cmake contains this
```cmake
if ("${CMAKE_C_IMPLICIT_INCLUDE_DIRECTORIES}" STREQUAL "")
  set(CMAKE_C_IMPLICIT_INCLUDE_DIRECTORIES "${EMSCRIPTEN_SYSROOT}/include")
endif()
if ("${CMAKE_CXX_IMPLICIT_INCLUDE_DIRECTORIES}" STREQUAL "")
  set(CMAKE_CXX_IMPLICIT_INCLUDE_DIRECTORIES "${EMSCRIPTEN_SYSROOT}/include")
endif()
```

**Comment by SirLynix** [comment]
Oh it's because it's an old version of emsdk, I will update it

**Comment by SirLynix** [comment]
```
/home/runner/.config/.xmake/cache/packages/2212/l/libsdl/2.26.1/source/build_a44fcf1b/include/SDL_stdinc.h:186:5: warning: declaration shadows a variable in the global scope [-Wshadow]
    SDL_FALSE = 0,
    ^
/home/runner/work/_temp/a7562b0b-e50a-4925-92a7-3b58f343f949/emsdk-main/upstream/emscripten/cache/sysroot/include/SDL/SDL_stdinc.h:117:5: note: previous declaration is here
    SDL_FALSE = 0,
    ^
```
There's a conflict between emscripten embedded libsdl and this package.

What I don't understand is why it happens.

See [official emscripten action](https://github.com/libsdl-org/SDL/actions/runs/3480348842/jobs/5820026147#step:6:86):
```
[79/425] /home/runner/work/_temp/5e8a11c6-ee2b-4cd3-bc70-5d5a1aa4aa99/emsdk-main/upstream/emscripten/emcc -DSDL_STATIC_LIB -DUSING_GENERATED_CONFIG_H -I/home/runner/work/SDL/SDL/build/include -I/home/runner/work/SDL/SDL/build/include-config-release -idirafter "/home/runner/work/SDL/SDL/src/video/khronos" -Werror  -Wall -fno-strict-aliasing -Werror=declaration-after-statement -Wdeclaration-after-statement -fvisibility=hidden -Wshadow -fdiagnostics-color=always  -DSDL_BUILD_MAJOR_VERSION=2 -DSDL_BUILD_MINOR_VERSION=25 -DSDL_BUILD_MICRO_VERSION=0 -O3 -DNDEBUG -Wno-warn-absolute-paths -MD -MT CMakeFiles/SDL2-static.dir/src/SDL.c.o -MF CMakeFiles/SDL2-static.dir/src/SDL.c.o.d -o CMakeFiles/SDL2-static.dir/src/SDL.c.o -c /home/runner/work/SDL/SDL/src/SDL.c
```

with the one from this package:
```
/home/runner/work/_temp/a7562b0b-e50a-4925-92a7-3b58f343f949/emsdk-main/upstream/emscripten/emcc -DSDL_STATIC_LIB -DUSING_GENERATED_CONFIG_H @CMakeFiles/SDL2-static.dir/includes_C.rsp -idirafter "/home/runner/.config/.xmake/cache/packages/2212/l/libsdl/2.26.1/source/src/video/khronos"  -Wall -fno-strict-aliasing -Wdeclaration-after-statement -fvisibility=hidden -Wshadow -fdiagnostics-color=always  -DSDL_BUILD_MAJOR_VERSION=2 -DSDL_BUILD_MINOR_VERSION=26 -DSDL_BUILD_MICRO_VERSION=1 -O3 -DNDEBUG -fPIC -Wno-warn-absolute-paths -MD -MT CMakeFiles/SDL2-static.dir/src/SDL.c.o -MF CMakeFiles/SDL2-static.dir/src/SDL.c.o.d -o CMakeFiles/SDL2-static.dir/src/SDL.c.o -c /home/runner/.config/.xmake/cache/packages/2212/l/libsdl/2.26.1/source/src/SDL.c
```

`-I` option should fix this as it should take priority on emscripten include dir.

**Comment by SirLynix** [comment]
```
/home/runner/work/xmake-repo/xmake-repo/aarch64-linux-musl-cross/bin/aarch64-linux-musl-gcc -DSDL_STATIC_LIB -DUSING_GENERATED_CONFIG_H -I/home/runner/.config/.xmake/cache/packages/2212/l/libsdl/2.26.1/source/build_e1d750e2/include -I/home/runner/.config/.xmake/cache/packages/2212/l/libsdl/2.26.1/source/build_e1d750e2/include-config-release -idirafter "/home/runner/.config/.xmake/cache/packages/2212/l/libsdl/2.26.1/source/src/video/khronos" -DHAVE_LINUX_VERSION_H  -Wall -fno-strict-aliasing -Wdeclaration-after-statement -fvisibility=hidden -Wshadow -fdiagnostics-color=always -I/usr/include -DMESA_EGL_NO_X11_HEADERS -DEGL_NO_X11 -D_REENTRANT  -DSDL_BUILD_MAJOR_VERSION=2 -DSDL_BUILD_MINOR_VERSION=26 -DSDL_BUILD_MICRO_VERSION=1 -O3 -DNDEBUG -MD -MT CMakeFiles/SDL2-static.dir/src/SDL_dataqueue.c.o -MF CMakeFiles/SDL2-static.dir/src/SDL_dataqueue.c.o.d -o CMakeFiles/SDL2-static.dir/src/SDL_dataqueue.c.o -c /home/runner/.config/.xmake/cache/packages/2212/l/libsdl/2.26.1/source/src/SDL_dataqueue.c
In file included from /home/runner/.config/.xmake/cache/packages/2212/l/libsdl/2.26.1/source/build_e1d750e2/include/SDL_stdinc.h:43,
                 from /home/runner/.config/.xmake/cache/packages/2212/l/libsdl/2.26.1/source/build_e1d750e2/include/SDL_assert.h:25,
                 from /home/runner/.config/.xmake/cache/packages/2212/l/libsdl/2.26.1/source/src/./SDL_internal.h:123,
                 from /home/runner/.config/.xmake/cache/packages/2212/l/libsdl/2.26.1/source/src/SDL.c:21:
/usr/include/stdio.h:27:10: fatal error: bits/libc-header-start.h: No such file or directory
   27 | #include <bits/libc-header-start.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

I'm not sure how to fix that, it looks like it includes the host system includes instead of the cross-compiler ones.

**Comment by SirLynix** [comment]
```
    set(X11_dirs)
    find_path(X_INCLUDEDIR
      NAMES X11/Xlib.h
      PATHS
        /usr/pkg/xorg/include
        /usr/X11R6/include
        /usr/X11R7/include
        /usr/local/include/X11
        /usr/include/X11
        /usr/openwin/include
        /usr/openwin/share/include
        /opt/graphics/OpenGL/include
        /opt/X11/include
    )

    if(X_INCLUDEDIR)
      list(APPEND EXTRA_CFLAGS "-I${X_INCLUDEDIR}")
      list(APPEND CMAKE_REQUIRED_INCLUDES ${X_INCLUDEDIR})
    endif()
```

I think the `-I/usr/include` comes from here, it looks like it finds X11 on the host system.

**Comment by waruqi** [comment]
> ```
>     set(X11_dirs)
>     find_path(X_INCLUDEDIR
>       NAMES X11/Xlib.h
>       PATHS
>         /usr/pkg/xorg/include
>         /usr/X11R6/include
>         /usr/X11R7/include
>         /usr/local/include/X11
>         /usr/include/X11
>         /usr/openwin/include
>         /usr/openwin/share/include
>         /opt/graphics/OpenGL/include
>         /opt/X11/include
>     )
> 
>     if(X_INCLUDEDIR)
>       list(APPEND EXTRA_CFLAGS "-I${X_INCLUDEDIR}")
>       list(APPEND CMAKE_REQUIRED_INCLUDES ${X_INCLUDEDIR})
>     endif()
> ```
> 
> I think the `-I/usr/include` comes from here, it looks like it finds X11 on the host system.

you can patch it or disable cross platform.

File: .github/workflows/cross_musl.yml [hunk 2]
```diff
@@ -30,5 +30,5 @@ jobs:
 
       - name: Tests
         run: |
-          xmake l ./scripts/test.lua -D -p cross --toolchain=muslcc --sdk=`pwd`/${{ matrix.cross }}-cross
+          xmake l ./scripts/test.lua -vD -p cross --toolchain=muslcc --sdk=`pwd`/${{ matrix.cross }}-cross
```
Comment (waruqi) [inline .github/workflows/cross_musl.yml:33]: it works now, please remove `-v`

Comment (SirLynix) [reply .github/workflows/cross_musl.yml:33]: I was wondering why this wasn't a default on xmake-repo CI, since it brings a lot of info

Comment (waruqi) [reply .github/workflows/cross_musl.yml]: Some packages output a lot of logs, making compilation very slow and difficult to locate critical error messages.

There is also a limit to the size of the logs displayed on the ci, resulting in me having to download the full logs to find the error messages.

**Pull request merged**